### ISSUE TITLE
`allunique(::String)` performance anomaly

### ISSUE THREAD

**Issue by chrstphrbrns** [JuliaLang/julia#50360]
It seems that `allunique` is not short-circuiting efficiently for strings
```julia
julia> VERSION
v"1.10.0-DEV.1604"

julia> using Random

julia> using BenchmarkTools

julia> s=randstring(100000);

julia> @btime allunique(s)
  140.850 μs (7 allocations: 848 bytes)
false

julia> @btime allunique(s[1:100]) && allunique(s)
  575.746 ns (8 allocations: 968 bytes)
false
```
```140 μs >> 575 ns```

**Comment by oscardssmith** [comment]
The performance issue is very silly. We are assuming calls to `length` are fast and therefore calculating the string's `length` 4 separate times. We probably need a specialized version of `allunique` for `String` that uses the number of codeunits instead.

**Comment by Jay-sanjay** [comment]
sir , are you still working on this issue because I want to work on it

**Comment by oscardssmith** [comment]
Go for it

**Comment by oscardssmith** [comment]
That is correct. I think the easiest fix would probably to add a special cased string version that uses `ncodeunits` instead.

**Comment by Jay-sanjay** [comment]
yeah , I was thinking to create a special version of 'allunique'  that would calculate the length of the string only once.

**Comment by Jay-sanjay** [comment]
@chrstphrbrns according to me this can be the implementation

julia> s = randstring(100000);

julia> @btime allunique(s)
  2.605 μs (4 allocations: 416 bytes)
false

julia> @btime allunique(s[1:100]) && allunique(s)
  256.077 ns (3 allocations: 112 bytes)
false

**Comment by oscardssmith** [comment]
The implementation here appears to be missing.

**Comment by Jay-sanjay** [comment]
function allunique(str::AbstractString)
    n = sizeof(str)
    codeunits = Vector{UInt32}(undef, n ÷ sizeof(UInt32))
    for (i, cu) in enumerate(eachindex(str))
        codeunits[i] = cu
    end

    sort!(codeunits)

    for i in 2:length(codeunits)
        codeunits[i - 1] == codeunits[i] && return false
    end

    return true
end

**Comment by Jay-sanjay** [comment]
Please review this one

**Comment by Jay-sanjay** [comment]
@LilithHafner is the issues solved by above code

**Comment by LilithHafner** [comment]
@Jay-sanjay, no. Running that code on te input "a" gives
```julia
julia> allunique("a")
ERROR: BoundsError: attempt to access 0-element Vector{UInt32} at index [1]
Stacktrace:
 [1] setindex!
   @ Main ./array.jl:1019 [inlined]
 [2] allunique(str::String)
   @ Main ./REPL[7]:5
 [3] top-level scope
   @ REPL[14]:1
```
I also prefer reviewing code when it is in a pull request rather than an issue because in a pull request, we automatically run CI tests to identify this sort of issue.

There is also already an open pull request for this issue (#50644). While it is certainly possible to have multiple PRs open for the same issue, it sometimes fragments reviewer effort, so it can often be more helpful to contribute to the existing PR.

**Comment by LilithHafner** [comment]
@chrstphrbrns How did you come across this issue? I have trouble imagining a use case for `allunique(::String)`.

**Comment by chrstphrbrns** [comment]
I found it while looking into `unique`/`allunique` performance. I agree it's not a high-priority optimization, but it looks like @bvdmitri fixed it

**Comment by Jay-sanjay** [comment]
sir I guess Some checks were not successful.

**Comment by bvdmitri** [comment]
I guess automatic checks are not very reliable and it often happens in many other PRs
